controllers/virtualmachinesnapshot/virtualmachinesnapshot_controller.go

```diff
@@ -234,21 +234,43 @@ func (r *Reconciler) ReconcileDelete(ctx *pkgctx.VirtualMachineSnapshotContext)
 		return errVMNameEmpty
 	}
 
+	// If CSI hasn't complated the sync and Snapshot is being deleted,
+	// we should notify all impacted SPUs to update their requested capacity.
+	// This should be a rare case.
+	// This has to be done before checking if VM exists, since VM deletion will
+	// only sync SPU with same StorageClass, not the SPUs for PVCs who might
+	// have different storageClasses.
+	if vmSnapshot.Annotations[constants.CSIVSphereVolumeSyncAnnotationKey] !=
+		constants.CSIVSphereVolumeSyncAnnotationValueCompleted &&
+		vmSnapshot.Status.Storage != nil &&
+		vmSnapshot.Status.Storage.Requested != nil {
+
+		for _, requested := range vmSnapshot.Status.Storage.Requested {
+			ctx.StorageClassesToSync.Insert(requested.StorageClass)
+		}
+	}
+
 	vm := &vmopv1.VirtualMachine{}
 	objKey := client.ObjectKey{Name: vmSnapshot.Spec.VMName, Namespace: vmSnapshot.Namespace}
 	if err := r.Get(ctx, objKey, vm); err != nil {
 		if apierrors.IsNotFound(err) {
 			ctx.Logger.V(5).Info("VirtualMachine not found, assuming the " +
 				"snapshot is deleted along with moVM, remove finalizer")
 			controllerutil.RemoveFinalizer(vmSnapshot, Finalizer)
-			// We don't sync SPU since we don't know which StorageClass of the VM.
+			// No need to explicitly enqueue SPU update here since VM deletion
+			// will trigger SPU sync for usage update.
 			return nil
 		}
 		return fmt.Errorf("failed to get VirtualMachine %q: %w", objKey, err)
 	}
 
-	// Enqueue the storage class to sync corresponding SPU.
-	ctx.StorageClassesToSync.Insert(vm.Spec.StorageClass)
+	// Only need to sync SPU with same SC as the VM since VMSnapshot SPU's
+	// used capacity excludes PVC's usage.
+	if vmSnapshot.Annotations[constants.CSIVSphereVolumeSyncAnnotationKey] ==
```

Comment:
Just wondering: this should happen for all SCs referenced by the snapshot so we can reduce (undo) the requested storage from the SPU for this SC.

Also, we should make sure this is called from ReconcileNormal as well to handle the same scenario.